Adds --columns for Print using custom width #13

sharifhsn opened the pull request:
This pull request fixes #13 partially by adding the `--columns` option to manually set the number of columns. The default is 2 as with previous behavior.

Some notes:
- the short option is `-w` not `-c` as that option is already reserved for `--bytes`
- it passes `cargo test`
- I could have added the option to automatically set columns to terminal width, but that would require adding another dependency

sharifhsn commented:
I have a pull request ready for terminal width, but it requires adding the `terminal_width` dependency. I'll wait to submit it until this request is approved.

sharkdp commented on src/bin/hexyl.rs:
```diff
@@ -149,6 +149,17 @@ fn run() -> Result<(), AnyhowError> {
+                .value_name("N")
+                .help(
+                    "Sets the number of hex data columns to be displayed. \
+                    Cannot be used with other width-setting options.",
```
Should this be omitted for now?

sharifhsn replied:
I already have the pull request for terminal width ready, I'll commit it if there's no other problems. I just wanted to make sure that the major architectural changes were correct.

sharkdp replied:
Ok, in that case, yes. Please add it to this PR. Or leave the help text as-is, and we do a second PR with the auto-width changes.

sharkdp commented on src/squeezer.rs:
```diff
@@ -111,13 +110,14 @@ impl Squeezer {
     use super::*;
 
-    const LSIZE_USIZE: usize = LSIZE as usize;
+    const LSIZE: u64 = 16;
+    const LSIZE_USIZE: usize = 16;
```
Maybe
```suggestion
    const LSIZE: u64 = 16;
    const LSIZE_USIZE: usize = LSIZE as usize;
```

sharkdp replied:
Or just use `LSIZE` and use `LSIZE as usize` in a few places below?

sharkdp reviewed (commented):
Thank you very much. Works great!

Does this have any impact on performance? (see e.g. https://github.com/sharkdp/hexyl/issues/66#issuecomment-511125330 on how to perform `hexyl` benchmarks using `hyperfine`)

sharifhsn commented:
> Thank you very much. Works great!
> 
> Does this have any impact on performance? (see e.g. [#66 (comment)](https://github.com/sharkdp/hexyl/issues/66#issuecomment-511125330) on how to perform `hexyl` benchmarks using `hyperfine`)

No, I'll do that now. I couldn't find a benchmark harness before, thank you for pointing me to it.

sharkdp commented:
We should probably add a script with a few useful benchmarks, true. For this PR, I would suggest to use something like:

```bash
#!/bin/bash

dd if=/dev/urandom bs=1M count=1 of=/tmp/data

hyperfine \
    --warmup 3 \
    --shell=none \
    --output=pipe \
    --export-markdown=result-pipe.md \
    'hexyl /tmp/data'

hyperfine \
    --warmup 3 \
    --shell=none \
    --output=inherit \
    --export-markdown=result-inherit.md \
    'hexyl /tmp/data'

rm /tmp/data
```

Here, we use `hyperfine`s new `--output` option to specify where to redirect the output of the benchmarked program. Using "pipe" is like doing `hexyl … | cat > /dev/null` (which is much different from doing `hexyl … > /dev/null` because `hexyl` doesn't directly write to `/dev/null` and kernel optimizations don't apply, i.e. `write` calls actually need to take place). And using "inherit" is like calling the program with the TTY attached (this will output a lot to your screen).

sharkdp commented:
You can also create two binaries, `hexyl-master` and `hexyl-feature`, and then run something like:
```
hyperfine \
    --warmup 3 \
    --shell=none \
    --output=pipe \
    --parameter-list version master,feature
    --export-markdown=result-pipe.md \
    './hexyl-{version} /tmp/data'
```
to see the direct comparison.

sharifhsn commented:
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `hexyl data` | 421.3 ± 14.3 | 404.5 | 444.0 | 1.24 ± 0.11 |
| `hexyl --no-squeezing data` | 563.5 ± 14.7 | 542.3 | 583.4 | 1.65 ± 0.15 |
| `target/release/hexyl data` | 433.0 ± 16.0 | 409.1 | 462.5 | 1.27 ± 0.12 |
| `target/release/hexyl --no-squeezing data` | 526.2 ± 18.1 | 506.6 | 559.6 | 1.55 ± 0.14 |
| `target/release/hexyl --columns 4 data` | 367.4 ± 12.7 | 347.8 | 382.6 | 1.08 ± 0.10 |
| `target/release/hexyl --columns 4 --no-squeezing data` | 407.6 ± 14.2 | 375.5 | 426.5 | 1.20 ± 0.11 |
| `target/release/hexyl --columns 8 data` | 340.5 ± 28.6 | 306.6 | 399.2 | 1.00 |
| `target/release/hexyl --columns 8 --no-squeezing data` | 351.9 ± 15.3 | 318.9 | 366.2 | 1.03 ± 0.10 |

These are my results using 1 MB of `/dev/urandom`. I used my system-installed binary of `hexyl` to compare since I'm pretty sure there were no feature changes since the last release.

It seems as though for the default width of two columns the new version is slightly slower, but within noise for the benchmark. With more columns it's much faster, but that of course makes sense since it's writing fewer lines.

sharifhsn commented:
One more thing; I've implemented `columns` here as a `u8`, since I find it unlikely that anyone would ever use over 255 columns in the viewer. This also places a cap on terminal width at 8969. Is this desirable, or should I change the type to `u16`?

sharifhsn commented:
I've added the terminal width setting, so this should completely close #13.

sharkdp commented:
> One more thing; I've implemented `columns` here as a `u8`, since I find it unlikely that anyone would ever use over 255 columns in the viewer. This also places a cap on terminal width at 8969. Is this desirable, or should I change the type to `u16`?

I wanted to remark this in my first review but didn't want to be overly annoying :smile:. In my personal experience: trying to use size-limited (unsigned) integers (u8, u16, i8, i16) in places which are *not* size- or performance critical is almost always a mistake. It's usually meant as a (premature) optimization. But the reality is that it (1) leads to more errors due to overflows (2) lead to more complicated code due to more explicit casts between 8-bit, 16-bit and 32-bit numbers and (3) don't even necessarily generate faster code (because if they are not closely packed, we can't really make use of them) or smaller data structs (due to alignment mismatches).

So personally, for a single parameter, I would never mind trying to go for a small integer type and simply go with a 32-bit integer, or even `isize`/`usize`. Often, it's even undesirable to use unsigned integers for parameters that cannot (reasonably) be negative, because again, it might lead to more complicated code, and it can lead to tricky overflow errors.

sharkdp commented:
Feel free to bump the min. supported Rust version if that causes problems.

sharkdp commented:
Concerning the newly added options:

Would it be possible to scratch `--auto-width` and use something like `--columns=max` or `--columns=auto` instead?

Also, I think `--terminal-width` should be usable with other "width-setting" options. I think of it purely as an overwrite for the auto-detected terminal width. What do you think?

> I've added the terminal width setting, so this should completely close #13.

Does it? I believe the issue also asks for custom widths that would not be multiples of 16. Like if you have some data structure with elements that are 14 byte long. Then you might want to specify `--width=14`.

sharifhsn commented:
> > One more thing; I've implemented `columns` here as a `u8`, since I find it unlikely that anyone would ever use over 255 columns in the viewer. This also places a cap on terminal width at 8969. Is this desirable, or should I change the type to `u16`?
> 
> I wanted to remark this in my first review but didn't want to be overly annoying smile. In my personal experience: trying to use size-limited (unsigned) integers (u8, u16, i8, i16) in places which are _not_ size- or performance critical is almost always a mistake. It's usually meant as a (premature) optimization. But the reality is that it (1) leads to more errors due to overflows (2) leads to more complicated code due to more explicit casts between 8-bit, 16-bit and 32-bit numbers and (3) doesn't even necessarily generate faster code (because if they are not closely packed, we can't really make use of them) or smaller data structs (due to alignment mismatches).
> 
> So personally, for a single parameter, I would never mind trying to go for a small integer type and simply go with a 32-bit integer, or even `isize`/`usize`. Often, it's even undesirable to use unsigned integers for parameters that cannot (reasonably) be negative, because again, it might lead to more complicated code, and it can lead to tricky overflow errors.

Thank you for saying this! I am quite new to Rust so I appreciate the advice. Making it a `u16` eliminates a cast (since `terminal_width` uses a `u16`) so I'll use that instead. Using a bigger value would be more of a hassle than a `u16`.

sharkdp commented on src/bin/hexyl.rs:
```diff
@@ -267,6 +305,50 @@ fn run() -> Result<(), AnyhowError> {
+        {
+            Some(terminal_width)
+        } else if matches.is_present("auto_width") {
+            Some(terminal_size().expect("not a tty").0 .0)
```
I think we should properly handle this case (not a tty), by which I mean: return a proper error here and show a nice error message for the user. This can easily be triggered if the output of hexyl is being redirected.

sharkdp commented on src/lib.rs:
```diff
@@ -164,6 +166,7 @@ impl<'a, Writer: Write> Printer<'a, Writer> {
         show_position_panel: bool,
         border_style: BorderStyle,
         use_squeeze: bool,
+        columns: u8,
```
This is a breaking change for `hexyl`-as-a-library. This whole `new` method is really unfortunate. There are too many arguments even before this PR. We should really use a builder-pattern instead.

But that's not really the problem of your PR.

sharifhsn replied:
I would be willing to work on implementing a builder pattern if you're committed to pushing a breaking change anyway.

sharkdp commented on src/lib.rs:
```diff
@@ -347,45 +351,57 @@ impl<'a, Writer: Write> Printer<'a, Writer> {
-                );
+            if len < usize::from(8 * self.columns) {
+                write!(&mut self.buffer_line, "{0:1$}", "", 3 * (8 - len % 8))?;
+                // dbg!(usize::from(8 * self.columns) - (len + (8 - len % 8)));
```
remove?

sharifhsn replied:
Ah, my sloppiness strikes again. Good eye.

sharifhsn commented:
> Concerning the newly added options:
> 
> Would it be possible to scratch `--auto-width` and use something like `--columns=max` or `--columns=auto` instead?

Of course.

> Also, I think `--terminal-width` should be usable with other "width-setting" options. I think of it purely as an overwrite for the auto-detected terminal width. What do you think?

I was only going off of the original issue's suggestions. The use case for `--terminal-width` is probably in a situation where the width is necessarily limited for some reason. In that case, I would pick the minimum between the two options i.e. prefer `--columns` unless the number of columns exceeds that which is dictated by `--terminal-width`.

> > I've added the terminal width setting, so this should completely close #13.
> 
> Does it? I believe the issue also asks for custom widths that would not be multiples of 16. Like if you have some data structure with elements that are 14 byte long. Then you might want to specify `--width=14`.

Ah, I wasn't aware that was in the scope of the original issue. That would probably require a more significant rewrite.

sharkdp commented:
> Ah, I wasn't aware that was in the scope of the original issue. That would probably require a more significant rewrite.

That's okay. Let's focus on `--columns` for now.

sharkdp commented:
Another option would be to leave the option name as is, but change the meaning. And then for this PR, we only allow `--columns=8,16,24,…`. What do you think?

sharifhsn commented:
> Another option would be to leave the option name as is, but change the meaning. And then for this PR, we only allow `--columns=8,16,24,…`. What do you think?

I think it would be more ergonomic here to have the two options, columns-per-panel and number-of-panels.

In my mind, the idea behind variable number columns is to have the panel size itself adjust. With the other option, the panel size would always be 8 bytes, and a future ability to have e.g. `columns=14` would leave blank space at the end.

It would be better to set `columns=7` and `panels=2` which is more explicit in my opinion.

sharkdp commented:
You are right.

Let's change the command line option name then 👍🏼

sharkdp commented on src/bin/hexyl.rs:
```diff
@@ -149,6 +151,33 @@ fn run() -> Result<(), AnyhowError> {
+        )
+        .arg(
+            Arg::new("terminal_width")
+                .short('t')
```
I'm trying to be conservative with short options. There are only a few of them :-). I would suggest that we remove both short options for now (i.e.: same for `-w`).

sharkdp commented:
> I have a commit ready for a builder pattern. It uses a new PrinterBuilder struct, so it can technically be implemented in a non-breaking way, although that's probably undesirable. It's quite out of scope for this pull request, so I'll submit a new one once this one is approved.

That would be great! If we do that, I would make that a breaking change, i.e. remove the old `new` method.

> @sharkdp I think we can declare this PR complete? I don't see any other roadblocks.

Yes. Looks great. Two things: the inline comment regarding the short options. And it would be great if you could add a CHANGELOG entry in the "unreleased" section. The format for entries is:
```
- Description what has been changed, see #123 (@user)
```
where `#123` links to the bug ticket and/or the PR and `user` is your username.

sharkdp commented:
Thank you!
